Code/Engine/ECS/Systems/PlayerControllerSystem.cpp

```diff
@@ -0,0 +1,116 @@
+	if (abs(input.GetLeftAxis("Vertical")) > input.ControllerDeadZone)
+	{
+		usingController = true;
+		transformComponent.m_Transform->Translate(glm::vec3(0.0f, 0.0f, -input.GetLeftAxis("Vertical") * movespeed));
+	}
+
+	if (input.GetKey(sad::KeyCode::W) && !usingController)
+	{
+		transformComponent.m_Transform->Translate(glm::vec3(0.0f, 0.0f, 1.0f * movespeed));
+	}
+
+	if (input.GetKey(sad::KeyCode::S) && !usingController)
+	{
+		transformComponent.m_Transform->Translate(glm::vec3(0.0f, 0.0f, -1.0f * movespeed));
+	}
+
+	// Handles left/right movement using A and D + controller left joystick.
+	if (abs(input.GetLeftAxis("Horizontal")) > input.ControllerDeadZone)
+	{
+		usingController = true;
+		transformComponent.m_Transform->Translate(glm::vec3(-input.GetLeftAxis("Horizontal") * movespeed, 0.0f, 0.0f));
+	}
+
+	if (input.GetKey(sad::KeyCode::A) && !usingController)
+	{
+		transformComponent.m_Transform->Translate(glm::vec3(1.0f * movespeed, 0.0f, 0.0f));
+	}
+
+	if (input.GetKey(sad::KeyCode::D) && !usingController)
+	{
+		transformComponent.m_Transform->Translate(glm::vec3(-1.0f * movespeed, 0.0f, 0.0f));
+	}
+
+	// Handles left/right rotation using LEFT and RIGHT arrow keys.
+	if (input.GetKey(sad::KeyCode::LeftArrow) || input.GetButton(sad::ControllerButton::L1))
+	{
+		transformComponent.m_Transform->Rotate(glm::vec3(0.0f, 1.0f * 0.25f, 0.0f));
+	}
+
+	if (input.GetKey(sad::KeyCode::RightArrow) || input.GetButton(sad::ControllerButton::R1))
+	{
+		transformComponent.m_Transform->Rotate(glm::vec3(0.0f, -1.0f * 0.25f, 0.0f));
+	}
+}
+
+void sad::PlayerControllerSystem::EditorControls(InputManager& input, const ecs::TransformComponent& transformComponent, float movespeed)
+{
+	// Handles up/down flight using SPACE and LSHIFT.
+	if (input.GetKey(sad::KeyCode::Space) || input.GetButton(sad::ControllerButton::South))
+	{
+		transformComponent.m_Transform->Translate(glm::vec3(0.0f, 1.0f * movespeed, 0.0f));
+	}
+
+	if (input.GetKey(sad::KeyCode::LShift) || input.GetButton(sad::ControllerButton::East))
+	{
+		transformComponent.m_Transform->Translate(glm::vec3(0.0f, -1.0f * movespeed, 0.0f));
+	}
+}
+
+void sad::PlayerControllerSystem::Update()
+{
+	InputManager& input = InputManager::GetInstance();
+
+	// Test mouse position
+	if (input.GetMouseButtonPressed(SDL_BUTTON_MIDDLE))
+	{
+		std::string mousePosition = "Mouse Pos: (";
+		mousePosition += std::to_string(input.GetMousePosition().x);
+		mousePosition += ", ";
+		mousePosition += std::to_string(input.GetMousePosition().y);
+		mousePosition += ")";
+
+		core::Log(ELogType::Info, mousePosition.c_str());
+	}
+
+	float movespeed = 0.005f;
+
+	sad::ecs::EntityWorld& world = sad::ecs::Registry::GetEntityWorld();
+	auto playerView = world.view<const sad::ecs::PlayerControllerComponent, const sad::ecs::TransformComponent>();
+	auto editorView = world.view<const sad::ecs::EditorControllerComponent, const sad::ecs::TransformComponent>();
```

Comment:
Not too sure about any performance differences here, but we might be able to make use of a cool entt feature called `exclude`.

In this case, we make the assumption that only one of these controllers should exist at a time - so we can exclude the opposite component in both view query (ie: exclude the `EditorControllerComponent` in our search for the `PlayerControllerComponent`). Might also be good just to be really explicit with the fact that their existence is mutual so we don't accidentally update both at once 😵 . 

An example from the entt crash course!

![image](https://user-images.githubusercontent.com/48817786/199865208-d32f0dbe-eef7-4be8-841a-a22bc3308384.png)